working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
 
-#### Record binder
+This pattern is particularly useful for writing code in an algebraic datatype
+style. For example:
 
-A record pattern destructures fields from records and objects.
+```dart
+class Rect {
+  final double width, height;
 
-```
-recordBinder        ::= '(' recordFieldBinders ')'
+  Rect(this.width, this.height);
+}
 
-recordFieldBinders  ::= recordFieldBinder ( ',' recordFieldBinder )* ','?
-recordFieldBinder   ::= ( identifier ':' )? binder
-                      | identifier ':'
+display(Object obj) {
+  switch (obj) {
+    case Rect(width: w, height: h): print('Rect $w x $h');
+    default: print(obj);
+  }
+}
 ```
 
-Each field is either a binder which destructures a positional field, or a binder
-prefixed with an identifier and `:` which destructures a named field.
-
-Each named field in the record pattern is matched by calling a corresponding
-getter on the matched object. A record pattern can be applied to an object of
-any type with any set of named field patterns as long as the object being
-matched has getters with those names. (Positional fields, however, can only be
-matched by record values.)
-
-When destructuring named fields, it's common to want to bind the resulting value
-to a variable with the same name. As a convenience, the binder can be omitted on
-a named field. In that case, the field implicitly contains a variable binder
-subpattern with the same name. These are equivalent:
+You can also use an extractor to both match an enum value and destructure
+fields from it:
 
 ```dart
-var (first: first, second: second) = (first: 1, second: 2);
-var (first:, second:) = (first: 1, second: 2);
-```
-
-**TODO: Allow a `...` element in order to ignore some positional fields while
-capturing the suffix.**
+enum Severity  {
+  error(1, "Error"),
+  warning(2, "Warning");
 
-#### Wildcard binder
+  Severity(this.level, this.prefix);
 
-A wildcard binder pattern does nothing.
+  final int level;
+  final String prefix;
+}
 
-```
-wildcardBinder ::= "_"
+log(Severity severity, String message) {
+  switch (severity) {
+    case Severity.error(prefix: p):
+      print('!! $p !! $message'.toUppercase());
+    case Severity.warning(prefix: p):
+      print('$p: $message');
+  }
+}
 ```
 
-It's useful in places where you need a subpattern in order to destructure later
-positional values:
+The `extractorName` must refer to a named type or an enum value. (If you want to
+use an extractor with a function type, you can use a typedef.)
 
-```
-var list = [1, 2, 3];
-var [_, two, _] = list;
-```
+It is a compile-time error if:
 
-#### Variable binder
+*   `extractorName` does not refer to a type or enum value.
 
-A variable binding pattern matches the value and binds it to a new variable.
-These often occur as subpatterns of a destructuring pattern in order to capture
-a destructured value.
+*   A type argument list is present and does not match the arity of the type of
+    `extractorName` or `extractorName` is an enum value.
```

Comment:
Ah, so enums patterns cannot have type arguments, but these days enum values can.
Seems wrong that you can't do `case MyEnum<final T>.thing(:T defaultValue)`.

(Still feels odd to destructure an enum to begin with, but at least allow it completely.)